bluemira/base/reactor_config.py

```diff
@@ -0,0 +1,234 @@
+    local_params: Dict
+
+
+_PfT = TypeVar("_PfT", bound=ParameterFrame)
+_PARAMETERS_KEY = "params"
+
+
+class ReactorConfig:
+    """
+    Class that provides a simple interface over config JSON files and
+    handles overwriting multiply defined attributes.
+
+    If an attribute is defined more than once in a component,
+    the more globally scoped value is used (global overwrites local).
+
+    Parameters
+    ----------
+    config_path: str | dict
+        The path to the config JSON file or a dict of the data.
+
+    Example
+    -------
+
+    .. code-block:: python
+
+        @dataclass
+        class GlobalParams(ParameterFrame):
+            a: Parameter[int]
+
+
+        reactor_config = ReactorConfig(
+            {
+                "params": {"a": 10},
+                "comp A": {
+                    "params": {"a": 5, "b": 5},
+                    "designer": {
+                        "params": {"a": 1},
+                        "some_config": "some_value",
+                    },
+                    "builder": {
+                        "params": {"b": 1, "c": 1},
+                        "another_config": "another_value",
+                    },
+                },
+                "comp B": {
+                    "params": {"b": 5},
+                    "builder": {
+                        "third_config": "third_value",
+                    },
+                },
+            },
+            GlobalParams,
+        )
+
+    """
+
+    @staticmethod
+    def _read_json_file(path: str) -> dict:
+        with open(path, "r") as f:
+            return json.load(f)
+
+    def __init__(
+        self,
+        config_path: Union[str, dict],
+        global_params_type: Type[_PfT],
+    ):
+        if isinstance(config_path, str):
+            self.config_data = ReactorConfig._read_json_file(config_path)
+        elif isinstance(config_path, dict):
+            self.config_data = config_path
+        else:
+            raise ReactorConfigError("Invalid config_path")
+
+        self.global_params: _PfT = make_parameter_frame(
+            self.config_data.get(_PARAMETERS_KEY, {}),
+            global_params_type,
+        )
+
+    def __str__(self) -> str:
+        """Returns config_data as a nicely pretty formatted string"""
+        return self._pprint_dict(self.config_data)
+
+    def _pprint_dict(self, d: dict) -> str:
+        return pprint.pformat(d, sort_dicts=False, indent=1)
+
+    def _warn_on_duplicate_keys(self, shared_key: str, arg: str):
+        bluemira_warn(f"'{shared_key}' is already defined for {arg}")
+
+    def _check_key_in(self, key: str, config_layer: dict):
+        if key not in config_layer:
+            raise ReactorConfigError(
+                f"'{key}' not present in config:\n{self._pprint_dict(config_layer)}",
+            )
+
+    def _check_args_are_strings(self, args):
+        for a in args:
+            if not isinstance(a, str):
+                raise ReactorConfigError("args must strings")
```

Comment:
I think a normal `TypeError` is probably more appropriate here

Reply:
True